src/Swarm/Game/Display.hs

```diff
@@ -128,9 +128,11 @@ instance ToJSON Display where
 
 -- | Look up the character that should be used for a display.
 displayChar :: Display -> Char
-displayChar disp = case disp ^. curOrientation of
-  Nothing -> disp ^. defaultChar
-  Just dir -> M.lookup dir (disp ^. orientationMap) ? (disp ^. defaultChar)
+displayChar disp = fromMaybe (disp ^. defaultChar) $ do
+  dir <- disp ^. curOrientation
+  case dir of
+    DAbsolute d -> M.lookup d (disp ^. orientationMap)
+    _ -> Nothing
```

Comment:
```suggestion
  DAbsolute d <- disp ^. curOrientation
  M.lookup d (disp ^. orientationMap)
```

I think this ought to work --- pattern-matching the result of a bind results in a call to `fail`, which in the case of the `Maybe` monad would yield `Nothing`, exactly what is wanted here.  I don't have a strong preference though, pick whichever you think is clearer.